### DISCUSSION TITLE
Autowiring lazily

### DISCUSSION THREAD

**Discussion by marcotama** [mapstruct/mapstruct#2852] [Q&A] [answered]
I have a large DTO with a bunch of objects with a web of dependencies to each other, so all my mappers "use" each other and there are a number of cycles. Is there a way to tell MapStruct to `@Autowire` the other mappers `@Lazy`ly? I am using `@Mapper(componentModel = "spring")`.

**Comment by Zegveld** [comment, marked as answer]
As far as I know this is currently not possible, however you could use the `field` injection strategy (default way) and spring should be able to handle this correctly.

According to what I've read, the `constructor` injection route would result in an exception, while the `field` injection does not.
Constructor injection flow --> spring tries to create A notices it requires B, tries to create B notices it requires A, throws exception.
Field injection flow --> spring creates A, and notices it requires B, it creates B and notices it requires A, injects the previously created A into B, and injects the previously creates B into the previously created A. No exception.

source: https://stackoverflow.com/questions/3485347/circular-dependency-in-spring

**Reply by marcotama** [reply]
That doesn't seem to work, regardless of whether I explicitly set `injectionStrategy = InjectionStrategy.FIELD` in the mappers (which should be by default anyway)

**Reply by marcotama** [reply]
I checked the generated `Impl` class, and it fields the mappers are autowired as fields, as I would expected. This is really strange.

**Reply by marcotama** [reply]
I got around the issue by using:
- `@Mapping(..., expression = "java(myMethod(...))")` on my mapper method,
- `@Autowired MyOtherMapper mapper` on the mapper,
- `public Target myMethod(Source obj) {...}` where `...` uses `mapper`

Not ideal, but it gets the job done without further issues.

**Reply by Zegveld** [reply]
I just checked and since Spring boot 2.6.0 circular references are restricted. You can allow them by changing a setting in Spring boot: `spring.main.allow-circular-references=true`.

Release notes: https://github.com/spring-projects/spring-boot/wiki/Spring-Boot-2.6-Release-Notes
